### PR TITLE
Maps chapter

### PR DESCRIPTION
**Pull request by hackeryarn**
I would love feedback on this and potentially including it a WIP chapter so I could get more feedback.

Resolves https://github.com/quii/learn-go-with-tests/issues/47

### CODE REVIEW

**Comment by quii** [comment]
Just a heads up, the build being broken is probably my fault :/ i cant get my IDE to play nice with markdown enough to make mdlint pass automatically. Will sort it soon

Thanks for the chapter, will give it a read at lunch!

**Comment by pityonline** [comment]
@quii I'll fix the markdownlint issues, so it won't interfere in this PR.

**Comment by hackeryarn** [comment]
Thanks @pityonline! I didn't want to put the fixes into this PR since they weren't related to the Maps chapter.

**Comment by gypsydave5** [comment]
Hey @hackeryarn - this is a really good start - you've definitely got something here, it's a good example and a nice size! Excellent job!

Here's some feedback - I've got two 'big' thoughts, and a few smaller ones. There are also a few typos but we can pick them afterwards.

## Big thoughts

>Another interesting property of Maps is that you can modify them without passing them as a pointer. This is because maps are a reference types. They don't actually hold any values. Instead, they point to the underlying data structure which houses the data.

I'm really glad you bring this up - but I think it needs a lot more explanation and exploring given that this is an early chapter and we've not really hit reference types yet. Take a look at [Go maps in action](https://blog.golang.org/go-maps-in-action) and write something about why `var dictionary map[string]string` won't work when you make a new map, and explain why you will see `dictionary = make(map[string]string)` and what `make` does. It need not be long - send the reader over to the Go docs for the finer details - but it's definitely worth talking about.

Secondly, when you say that the Map keys need to be comparable, I think you should make it clear _why_ they need to be comparable. Spend a bit of time explaining what comparable means and why this is 'key' to being a key. I think this is more important than listing the comparable types.

## Small thoughts

These are just some minor things probably just slipped your mind!

You start off with the following
```go
func TestSearch(t *testing.T) {
    dict := Dict{"test": "this is just a test"}

// snip!
}
```

And then say it fails with

```sh
./dict_test.go:8:9: undefined: Search
```

But this should fail because `Dict` isn't defined. It looks a bit like you've done what I always do - copy a later version of the code into an earlier part of the chapter - I think you introduce `Dict` a few lines further down. Review the code examples in the first bit of the chapter and make them match up with the code in the repo (which looks good to me! 😄) 

---

Consider being more verbose with your names - `Dictionary` instead of `Dict` etc

---

> We switched to using a Dict struct which we have not defined yet

Not sure this is technically a struct when it's a typealiased Map.

---

> Your test should now fail with a clear error message
> `dict_test.go:12: got want this is just a test given, test`

I don't think this is a very clear error message! I'd use the 'refactored' output from a bit further down for the assertion at the beginning. I'm happy you're refactoring the test to a helper function, but I think you should start with the test output looking good.

---

> We can greatly improve our module's usage by aliasing the Map and making Search a method.

This isn't really a module (in the Golang sense of the word) - just call it a dictionary for now!

---
```
var NotFoundError = errors.New("could not find the word you were looking for")
```

> We can get rid of the magic error in our Search function by bringing it up into a constant.

It's not a constant (yet!), it's a variable 😄 , so I'd reword the sentence to something involving 'extracting out a variable'

---

> This algorithm looks almost identical to Add except we switched when we update
This isn't really an algorithm - it's just a function!

I hope all of that made sense - let me know if it didn't, and let me or @quii know if you need a hand or any more feedback.

Really looking forward to adding this chapter!

**Comment by quii** [comment]
Thanks for that feedback @gypsydave5 👍

**Comment by hackeryarn** [comment]
@gypsydave5 Thank you so much for the feedback. It was very helpful. I think I covered most of your concerns.

I am conflicted on using `Dictionary` over `Dict` since `Dict` is a pretty common abbreviation. However, I can go either way on this one.

**Comment by quii** [comment]
Hey

This looks like good progress and i think goes over the concepts we'd want to whilst remaining relatively terse.

## Some specific feedback

### 1
https://github.com/hackeryarn/learn-go-with-tests/blob/d8b4c0a6854dacc79cceff88eb1e980463b060ee/maps.md#refactor

On the first refactor you jumped into the map being a `Dict`. Maybe just keep it as a `map` in the first code snippet. In your next bit "using a type alias" you then explain the refactor. So I think that would be less confusing. 

You also seem to have two "Using a type alias" sections one after another repeated :) 

### 2

Rather than 

```go
        if got == nil {
            t.Error("expected to receive and error.")
        } else {
            assertStrings(t, got.Error(), want)
        }
```

It is a little simpler instead to do

```go
if got == nil {
    t.Fatal("expected to get an error")
}

assertString(etc...
```

Typo: You expect to receive _an_ error, not _and_ error

I'd be _tempted_ to call `got` `err` here. I know the code comments things says you should call it `got` but i think in the case of two things being returned it's better to be clear that we are talking about errors.

### 3 reference types section

> This is because maps are a reference types

Change to 

> This is because `map` is a  reference type

> They don't actually hold any values. Instead, they point to the underlying data structure which houses the data.

_Technically_ they do hold data, just a reference to the underlying data structure

> Maps being a pointer is really good, because it means that no matter how big your Map gets. Go will only make one copy of it unless you force Go to make a copy.

You have an erroneous full stop in the middle there. I'm not sure if this reads _quite_ right, maps arent really "a pointer". Will try and think of a better way of phrasing that.

> These approaches, both create an empty hash map and point dict at it. Which ensures that you will never get a nil pointer exception.

Dont need the first comma. You've introduced the concept of `hash map` all of a sudden. You might want to explain that, or just call it a map. 

### 4 Note on declaring a new error

>  For example, if you are running a website. You might not want the user to see ErrNotFound, but instead redirect them to a add page.

Think you want to put a semi-colon instead of a period after website. The sentence seems to stop abruptly.

### general 

I'd rename `assertDef` to `definition`
And all `def` to `definition` :) 
I think i'd go for `Dictionary` instead of `Dict`. The main reason is to help our Chinese friends who are translating this. I'm not sure if `Dict` would be a well-understood type for them (@pityonline ?)

Hope that helps! It looks a lot but nothing major.

**Comment by hackeryarn** [comment]
@quii Thanks for all the suggestions! I made all the edits you asked for except for one.

I did not change the below line. It came after my explanation of `hash map`. Therefore, I don't think it needs another explanation.

> These approaches both create an empty hash map and point dict at it. Which ensures that you will never get a nil pointer exception.

I also need to finish updating the code samples to use `definition` before this is merged. I just wanted you to see the edits of the chapter.

**Comment by quii** [comment]
Ok LGTM! 

Nice one @hackeryarn

**Comment by quii** [comment]
Just pushed a commit to add it to the TOC. I've put it a little later than initially discussed as you've added methods and errors so it felt right to put them after the respective sections on those topics

**Comment by quii** [comment]
One more thing @hackeryarn are you on twitter? On every new chapter I like to spam my poor followers, would like to give you credit :)

**Comment by hackeryarn** [comment]
@quii That's where the chapter placement fits after those additions. I'm glad I could help with this amazing effort :smile: 

And yes I am on twitter same handle and github. https://twitter.com/hackeryarn
